text/3470-crabi-v1.md

```diff
@@ -0,0 +1,661 @@
+  identically.
+
+  Would an approach based on "type transformers" help here?
+
+- The handling of `Option` and `Result` seems similarly unfortunate. We have to
+  use a distinct type to avoid freezing the layout of these Rust types for all
+  time. But could we do better, somehow? Would an approach based on "type
+  transformers" help here?
+
+- Should we provide additional guaranteed niche optimizations for `Option` in
+  crABI 1.0? In order to keep new minor revisions of crABI compatible with
+  crABI 1.0, any existing type that doesn't have a niche optimization in crABI
+  1.0 can never switch to having a niche optimization until a new
+  (incompatible) major version, and ideally we want new crABI major versions to
+  be rare.
+
+  That said, we should probably keep *most* uses of niche optimizations rare,
+  as they require special-case handling in other languages.
+
+- Similarly, should we provide additional guaranteed niche optimizations for
+  `Result`? Which ones? The same issue applies: any optimizations we want to
+  provide for `Result` on existing types should be in crABI 1.0.
+
+- Should we provide *fewer* niche optimizations? Those for `NonZero` and
+  reference types provide obvious value; are those for `bool` and `char` really
+  useful enough to justify the special case in languages that will have to
+  handle them explicitly rather than automatically?
+
+- crABI specifies supports owned pointers via `Box`, but does not specify how
+  to free such objects other than passing them back to the caller. Should we
+  represent this in the type system explicitly, such as via a special allocator
+  (e.g. `NoDeallocate` to require passing back to the module that allocated it,
+  or an FFI deallocator parameterized with a `free` function in the type)? Or a
+  `ManuallyDrop`? Or should we use a normal `Box`?
+
+- Should crABI support passing owned values that have a non-trivial `Drop`
+  implementation, on the assumption that the API will specify that they must be
+  passed back to another exported function for freeing (which could call
+  `Drop`)? Or should crABI just prohibit types with non-trivial `Drop`
+  implementations?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- crABI makes it easier to define safer interfaces, making it more likely that
+  a library can define an interface that doesn't necessarily require `unsafe`
+  to call. Thus, crABI would benefit from a mechanism to allow specifying an
+  `extern "crabi"` function as explicitly safe, and then not requiring an
+  `unsafe` block to call that function. This RFC does not specify such a
+  mechanism, but there have been various proposals for such mechanisms for
+  `extern "C"`, and those mechanisms should work equivalently for `extern
+  "crabi"`.
+
+- Once `extern "C"` supports C-compatible handling of `u128` and `i128`,
+  `extern "crabi"` should do the same.
+
+- Extensible enums. To define types that allow for extension, crABI would
```

Comment:
Related to this, there are probably situations where it would be useful to declare a size and/or alignment of an enum to be larger than necessary, for future compatibility  to allow adding new data that would otherwise change the ABI.

Reply:
Padding fields are not ideal either as they can end up in JSON or toString representations too easily where they are noise at best and a bug at worst.

Reply:
Maybe the same issue: we have found it difficult to safely define a Rust type corresponding to a C or C++ `enum`, because an out-of-range representation of a Rust enum is immediate UB, unlike in C/C++.

Example: https://github.com/zcash/zcash/blob/2112e467ee31ea95cf81904a6aae397fa3d031ae/src/rust/src/zip339_ffi.rs#L11-L37

and the corresponding C/C++ type:
https://github.com/zcash/zcash/blob/2112e467ee31ea95cf81904a6aae397fa3d031ae/src/rust/include/rust/zip339.h#L17-L38